lib/spack/spack/cmd/solve.py

```diff
@@ -102,21 +104,29 @@ def solve(parser, args):
 
     specs = spack.cmd.parse_specs(args.specs)
 
-    # dump generated ASP program
-    result = asp.solve(
-        specs, dump=dump, models=models, timers=args.timers, stats=args.stats,
-        reuse=args.reuse,
+    # set up solver parameters
+    # Note: reuse and other concretizer prefs are passed as configuration
+    solver = asp.Solver()
+    output = sys.stdout if "asp" in show else None
+    result = solver.solve(
+        specs,
+        out=output,
+        models=models,
+        timers=args.timers,
+        stats=args.stats,
+        setup_only=not any(s in show for s in ("output", "solutions"))
     )
-    if 'solutions' not in dump:
+    if 'solutions' not in show:
         return
 
     # die if no solution was found
     result.raise_if_unsat()
 
-    # dump the solutions as concretized specs
-    if 'solutions' in dump:
+    # show the solutions as concretized specs
+    if 'solutions' in show:
```

Comment:
`opt` won't work without solutions still, but that can be a separate PR to fix it.